File: backend/app/api/v1/endpoints/notes.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from app.database import get_db
from app.models.note import Note
from app.models.user import User
from app.schemas.note import NoteCreate, NoteUpdate, NoteResponse
from app.auth import get_current_user

router = APIRouter()

@router.get("/", response_model=List[NoteResponse])
def get_notes(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    notes = db.query(Note).filter(Note.user_id == current_user.id).order_by(Note.updated_at.desc(), Note.created_at.desc()).offset(skip).limit(limit).all()
    return notes

@router.post("/", response_model=NoteResponse, status_code=status.HTTP_201_CREATED)
def create_note(
    note: NoteCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    db_note = Note(
        user_id=current_user.id,
        title=note.title,
        content=note.content,
        topic_id=note.topic_id
    )
    db.add(db_note)
    db.commit()
    db.refresh(db_note)
    return db_note

@router.patch("/{note_id}", response_model=NoteResponse)
def update_note(
    note_id: int,
    note_update: NoteUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    db_note = db.query(Note).filter(Note.id == note_id, Note.user_id == current_user.id).first()
    if not db_note:
        raise HTTPException(status_code=404, detail="Note not found")
        
    update_data = note_update.dict(exclude_unset=True)
    for key, value in update_data.items():
        setattr(db_note, key, value)
        
    db.commit()
    db.refresh(db_note)
    return db_note

@router.delete("/{note_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_note(
    note_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    db_note = db.query(Note).filter(Note.id == note_id, Note.user_id == current_user.id).first()
    if not db_note:
        raise HTTPException(status_code=404, detail="Note not found")
        
    db.delete(db_note)
    db.commit()
    return None

File: frontend/src/api/notes.ts
import api from './client';

export interface Note {
  id: number;
  user_id: number;
  title: string;
  content: string;
  topic_id?: number;
  created_at: string;
  updated_at?: string;
}

export const notesAPI = {
  getNotes: async (): Promise<Note[]> => {
    const response = await api.get('/notes/');
    return response.data as any;
  },
  createNote: async (data: { title: string, content?: string, topic_id?: number }): Promise<Note> => {
    const response = await api.post('/notes/', data);
    return response.data as any;
  },
  updateNote: async (id: number, data: { title?: string, content?: string, topic_id?: number }): Promise<Note> => {
    const response = await api.patch(`/notes/${id}`, data);
    return response.data as any;
  },
  deleteNote: async (id: number): Promise<void> => {
    await api.delete(`/notes/${id}`);
  }
};
